### PR TITLE
Add SingleImageRandomDotStereogramsOp to Contrib

### PR DESCRIPTION
**Pull request by Mazecreator**
This add an Op to convert 3D data into a 2D image SIRDS (Single Image Random Dot Stereogram) for scientific data review.

It is related to this concussion:
https://github.com/tensorflow/tensorflow/issues/8022

### CODE REVIEW

File: tensorflow/contrib/image/BUILD [hunk 1]
```diff
@@ -71,6 +71,41 @@ cuda_py_test(
     ],
 )
 
+
```
Comment (martinwicke) [inline tensorflow/contrib/image/BUILD]: remove two extra blank lines

File: tensorflow/contrib/image/BUILD [hunk 2]
```diff
@@ -81,4 +116,4 @@ filegroup(
         ],
     ),
     visibility = ["//tensorflow:__subpackages__"],
-)
+)
```
Comment (martinwicke) [inline tensorflow/contrib/image/BUILD]: Add final \n

File: tensorflow/contrib/image/python/ops/single_image_random_dot_stereograms.py [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+      Size of "DATA" window, must be equal to or smaller than output_image_shape, will be centered
+      and use convergence_dots_size for best fit to avoid overlap if possible
+    name: A name for the operation (optional).
+
+  Returns:
+    A `Tensor` of type `uint8`.
+    returns a Tensor of size output_image_shape with depth_values encoded into image
+  """
+  result = _sirds_ops.single_image_random_dot_stereograms( depth_values=depth_values,
+                                hidden_surface_removal=hidden_surface_removal,
```
Comment (martinwicke) [inline tensorflow/contrib/image/python/ops/single_image_random_dot_stereograms.py]: 4 space added indent after opening paren

File: tensorflow/contrib/image/python/ops/single_image_random_dot_stereograms.py [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+                                        dots_per_inch=None,
+                                        eye_separation=None, mu=None,
+                                        normalize=None, normalize_max=None,
+                                        normalize_min=None,
+                                        boarder_level=None,
+                                        number_colors=None,
+                                        generation_mode=None,
+                                        output_image_shape=None,
+                                        output_data_window=None):
+  r"""Output a RandomDotStereogram Tensor of shape "output_image_shape" for export via encode_PNG or encode_JPG OP.
```
Comment (martinwicke) [inline tensorflow/contrib/image/python/ops/single_image_random_dot_stereograms.py]: brief too long. <80chars per line

File: tensorflow/contrib/image/ops/single_image_random_dot_stereograms_ops.cc [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+
+out = sirds.eval()
+
+png = tf.image.encode_png(out).eval()
+
+with open('picture_out.png', 'wb') as f:
+    f.write(png)
+
+
+depth_values:           Z values of data to encode into "output_data_window" window, lower further away {0.0 floor(far), 1.0 ceiling(near) after normalization}, must be rank 2
```
Comment (martinwicke) [inline tensorflow/contrib/image/ops/single_image_random_dot_stereograms_ops.cc]: Can you look at https://www.tensorflow.org/community/documentation, there's a bit of special formatting for the docstring that makes the op documentation look good.

File: tensorflow/contrib/image/ops/single_image_random_dot_stereograms_ops.cc [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+    .Attr("normalize: bool = true")
+    .Attr("normalize_max: float = -100.0")
+    .Attr("normalize_min: float = 100.0")
+    .Attr("boarder_level: float = 0.0")
+    .Attr("number_colors: int = 256")    
+    .Attr("generation_mode: string = 'SIRDS'")
+    .Attr("output_image_shape: shape = { dim {size:1024} dim {size: 768} dim {size: 1}}")
+    .Attr("output_data_window: shape = { dim {size:1022} dim {size: 757}}")
+    // TODO(Mazecreator):  Add "SetShapeFN" to reflect proper output shape & Rank 2 error checking on Input
+    // TODO(Mazecreator):  Add better error handling for output shapes / windows settings
```
Comment (martinwicke) [inline tensorflow/contrib/image/ops/single_image_random_dot_stereograms_ops.cc]: Can you address these TODOs?

**Review by martinwicke — COMMENTED** [review]
Can you run clang-format -style=google on your C++ code, and buildifier on your BUILD file changes?

**Comment by martinwicke** [comment]
Jenkins, test this please.

**Comment by Mazecreator** [comment]
I have installed and run the google formatting tools as well as cleaned up the MD and docstring.

**Comment by Mazecreator** [comment]
@martinwicke ,

The changes look fine to me, is there anything I need to do?  I think the CLA issue is due to your edits?

**Review by RealTimeDeployment — APPROVED** [review]
Looks good sorry for taking so long and thank you!

**Review by martinwicke — CHANGES_REQUESTED** [review]

**Comment by martinwicke** [comment]
Yeah, don't worry about CLAbot, it gets confused.

**Comment by Mazecreator** [comment]
@martinwicke Let me fix what you have indicated above, I would like to keep the "generation_mode" in the params if that is okay, I can stop the attribute command and simply ignore the value of this attribute.  I think this will be handy in the future.

Do I need to issue a git command to pull back your updates or can I just update my current repository and pull again with the updates or does this loose your fixes?

**Comment by RealTimeDeployment** [comment]
Go ahead and update current and fix the things i missed as you need to. It actually pulled my main file, but i have everything wrote down. I added a few specs extra for VM for tensorflow & cudnn to use than i really need. Was trying to visualize the behaviour. Thought it would be useful support from some of the comments. Stereograms looks good. No need for border -> border. Doubles was extra  not needed. Wanted it to be an asset for the future. Are there any other offsets that need fixing? I think most everything was accounted for except a few fft maybe to deal with n%. Probably need to know how to reverse the sequence to correct anythimg still being offset. Probably just a few more calculations to complete the math lib, but you may have already account for that issue. But think this pr will be good. Std:: should already cover everything else. Let me know if there are any other pr i need to close out or any questions. Dont think i opened any other pr myself. Lost track as this ended up covering a lot of more than i expected. Thank You for allowing me to be a part of this.

**Comment by Mazecreator** [comment]
@martinwicke ,

I have addressed the latest change requests.  Let me know if there is anything else (or something I broke) that needs updating.

I did leave the 'generation_mode' parameter to define the future interface, but the Op will ignore any values passed for now.

I have also updated the demo page, if needed, I can create a brief version for the tensorflow tutorials.
https://github.com/Mazecreator/TensorFlow-SIRDS/blob/master/single_image_random_dot_stereograms/SIRDS%20Demo.ipynb

Thanks for the help and patients, I haven't used git before and it sometimes does things a little faster than expected.  Also, it seems like you have edited the docs to ~78 character line lengths, I am not sure this is documented [here](https://www.tensorflow.org/community/documentation), it might be worth adding this formatting restriction as I only saw reference to "short" not a line length for the different document parts.

**Comment by RealTimeDeployment** [comment]
Ignore the read me update if you wish. I shortened it to be less explicit. That way you can choose what to pass. I don't know if it is an issue you or not. Let me know if I need to revert it.

**Comment by Mazecreator** [comment]
I don't see any update, but if you are talking about removing the specifics in the "generation_mode" for SIRDS, I am okay with that change.

**Comment by RealTimeDeployment** [comment]
Just noticed one thing that was mentioned. I am curious if 1d was over looked as they are not listed seperately from the string arg. I didnt think about it before. 1d attributes? An oversight on my part.  Unneccessary ?  sirds returns values for 1d attributes. Already fixed @ lines 243 - 244 ?

**Comment by Mazecreator** [comment]
@RealTimeDeployment I am getting a little lost with your questions, is this @martinwicke 's repository?  I am not sure the connection to this review.  I see the lines 243 - 244 are STD::ROUND statements and not sure about the 1D reference and strings?   I am sure you can put in a [1,X] shape but haven't testing it for a 1D display, but then in my mind it would be clearer to use a simple X,Y plot rather than a SIRDS for evaluation.

**Comment by martinwicke** [comment]
@Mazecreator The updates you make to this branch of your repository will be reflected in this PR. 

I would like you to remove the unused arguments. Arguments are easy to add, but hard to remove, so we should be careful about adding them. And please do fix the boarder issue.

**Comment by Mazecreator** [comment]
@martinwicke all your update requests have been completed.

File: tensorflow/contrib/image/python/ops/single_image_random_dot_stereograms.py [hunk 7]
```diff
@@ -34,8 +34,6 @@ def single_image_random_dot_stereograms(
     normalize=None, normalize_max=None,
     normalize_min=None,
     border_level=None,
-    number_colors=None,
```
Comment (martinwicke) [inline tensorflow/contrib/image/python/ops/single_image_random_dot_stereograms.py]: You removed this option, why?

Comment (Mazecreator) [inline tensorflow/contrib/image/python/ops/single_image_random_dot_stereograms.py]: This was an error, I don't understand how it compiled and tested properly.  Maybe the older version of the PY file was held by Python Notebook.  It is fixed in the next commit.

Sorry, good catch.

**Review by martinwicke — APPROVED** [review]

**Comment by martinwicke** [comment]
Jenkins, test this please!

**Comment by martinwicke** [comment]
Jenkins, test this please.

**Comment by drpngx** [comment]
this makes the image ops test fail:
```
FAIL: //tensorflow/contrib/image:image_ops_test (see /var/lib/jenkins/workspace/tensorflow-pull-requests-cpu-python3/bazel-ci_build-cache/.cache/bazel/_bazel_jenkins/eab0d61a99b6696edb3d2aff87b585e8/execroot/workspace/bazel-out/local-py3-opt/testlogs/tensorflow/contrib/image/image_ops_test/test.log).
INFO: From Testing //tensorflow/contrib/image:image_ops_test:
==================== Test output for //tensorflow/contrib/image:image_ops_test:
Traceback (most recent call last):
  File "/var/lib/jenkins/workspace/tensorflow-pull-requests-cpu-python3/bazel-ci_build-cache/.cache/bazel/_bazel_jenkins/eab0d61a99b6696edb3d2aff87b585e8/execroot/workspace/bazel-out/local-py3-opt/bin/tensorflow/contrib/image/image_ops_test.runfiles/org_tensorflow/tensorflow/contrib/image/python/kernel_tests/image_ops_test.py", line 23, in <module>
    from tensorflow.contrib.image.python.ops import image_ops
  File "/var/lib/jenkins/workspace/tensorflow-pull-requests-cpu-python3/bazel-ci_build-cache/.cache/bazel/_bazel_jenkins/eab0d61a99b6696edb3d2aff87b585e8/execroot/workspace/bazel-out/local-py3-opt/bin/tensorflow/contrib/image/image_ops_test.runfiles/org_tensorflow/tensorflow/contrib/image/__init__.py", line 39, in <module>
    from tensorflow.contrib.image.python.ops.single_image_random_dot_stereograms import single_image_random_dot_stereograms
ImportError: No module named 'tensorflow.contrib.image.python.ops.single_image_random_dot_stereograms'
```

**Comment by Mazecreator** [comment]
//tensorflow/contrib/image:image_ops_test was an existing Op & Test in the contrib/image directory.  I am not sure if it is supported, bu this shouldn't be related to my addition of the SIRDS Op.

**Comment by drpngx** [comment]
If you read the error message, it does say:
```
ImportError: No module named 'tensorflow.contrib.image.python.ops.single_image_random_dot_stereograms'
```

**Comment by Mazecreator** [comment]
Hmmm... Okay, I will need help, I don't understand how the Testing works.  I didn't touch that test and don't  see where it references "single_image_random_dot_stereograms".  I seem to be able to compile the source locally and run the SIRDS, so not sure what would cause this problem.

**Comment by drpngx** [comment]
You can run tests yourself by using `bazel build --config=opt //tensorflow/contrib/image/...`. In that case, you have modified `__init__.py`, but did not change the build rule: http://google3/third_party/tensorflow/contrib/image/BUILD?l=73
You need to add `":single_image_random_dot_stereograms"` to the dependencies.

**Comment by Mazecreator** [comment]
Hi @drpngx ,

Sorry, I added the dependency but now I am in deeper.  I ran this command as it seems to run the tests:
bazel test --config=opt --config=cuda //tensorflow/contrib/image/...
I even tried this without cuda & opt, but I am getting a different error now.

    ERROR: /home/greg/tensorflow/tensorflow/contrib/image/BUILD:59:1: in deps attribute of py_test rule //tensorflow/contrib/image:image_ops_test: rule '//tensorflow/contrib/image:single_image_random_dot_stereograms' does not exist. Since this rule was created by the macro 'cuda_py_test', the error might have been caused by the macro implementation in /home/greg/tensorflow/tensorflow/tensorflow.bzl:863:12.
    ERROR: Analysis of target '//tensorflow/contrib/image:image_ops_test' failed; build aborted.
    INFO: Elapsed time: 2.839s
    ERROR: Couldn't start the build. Unable to run tests.

Sorry, I just don't understand what is going on under the hood with the tests.  I rand the command you had put in the message, but that only seemed to build and not run the tests.

**Comment by drpngx** [comment]
Yes, you are right, you need `bazel test`. Also, the correct dependency is `":single_image_random_dot_stereograms_py"`, not `":single_image_random_dot_stereograms"` as I had previously indicated.

**Comment by drpngx** [comment]
Jenkins, test this please.

**Comment by drpngx** [comment]
Could you possibly rebase instead of merge? That confuses googlebot. Thanks!

**Comment by Mazecreator** [comment]
Hi @drpngx ,

Thank you, that took care of the problem.  Now it tests okay.  Not sure why it had to be added or where to find that information, might be good to capture documentation like this for the Add Op help.  seems like indents are critical as well.

**Comment by Mazecreator** [comment]
@drpngx ,

I am not sure how to do that, should I pull a new tensorflow/master and add these files back in or can I somehow rebase my mazecreator/master?

I am new to git as well...

Sorry for the hassle.

**Comment by martinwicke** [comment]
@drpngx the CLAbot is confused because I fixed some indent in order to get past sanity. It can be ignored. All commits are from me or @mazecreator and we are in CLA. 

I'm assuming you'll squash it anyway so it's good to go I think.

**Comment by drpngx** [comment]
Sounds good. Thaanks!

**Pull request merged**